File: app/src/main/java/ru/santaev/clipboardtranslator/viewmodel/TranslateViewModel.java
package ru.santaev.clipboardtranslator.viewmodel;


import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;
import android.os.Handler;

import java.util.concurrent.TimeUnit;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import ru.santaev.clipboardtranslator.db.entity.Language;
import ru.santaev.clipboardtranslator.model.IDataModel;
import ru.santaev.clipboardtranslator.model.TranslateDirectionProvider;
import ru.santaev.clipboardtranslator.util.settings.AppPreference;

public class TranslateViewModel extends ViewModel{

    private static final long TRANSLATE_DELAY_MILLIS = TimeUnit.MILLISECONDS.toMillis(500);

    private MutableLiveData<String> translatedText = new MutableLiveData<>();
    private MutableLiveData<Boolean> progress = new MutableLiveData<>();
    private MutableLiveData<Boolean> failed = new MutableLiveData<>();

    private MutableLiveData<Language> originLang = new MutableLiveData<>();
    private MutableLiveData<Language> targetLang = new MutableLiveData<>();

    private Disposable disposable;
    private String originText;
    private Handler handler;
    private Runnable runTranslate;

    private AppPreference appPreference;
    private TranslateDirectionProvider translateDirectionProvider;

    private IDataModel dataModel;

    public TranslateViewModel(IDataModel dataModel) {
        this.dataModel = dataModel;
        translatedText.setValue("");
        progress.setValue(false);
        failed.setValue(false);
        appPreference = AppPreference.getInstance();
        translateDirectionProvider = new TranslateDirectionProvider();

        originLang.setValue(appPreference.getOriginLang());
        targetLang.setValue(appPreference.getTargetLang());
        originText = "";

        handler = new Handler();
    }

    public void onOriginTextChanged(String text){
        originText = text;
        translate();
    }

    public boolean onOriginLangSelected(Language lang){
        Language oldLang = originLang.getValue();
        originLang.setValue(lang);
        appPreference.setOriginLang(lang);

        boolean support = translateDirectionProvider.isSupportTranslate(lang, targetLang.getValue());

        if (lang != oldLang && support){
            translate();
        }
        return support;
    }

    public boolean onTargetLangSelected(Language lang){
        Language oldLang = targetLang.getValue();
        targetLang.setValue(lang);
        appPreference.setTargetLang(lang);

        boolean support = translateDirectionProvider.isSupportTranslate(originLang.getValue(), lang);

        if (lang != oldLang && support){
            translate();
        }
        return support;
    }

    public void onClickRetry() {
        translate();
    }

    public LiveData<String> getTranslatedText(){
        return translatedText;
    }

    public LiveData<Boolean> getProgress(){
        return progress;
    }

    public MutableLiveData<Boolean> getFailed() {
        return failed;
    }

    public LiveData<Language> getOriginLang() {
        return originLang;
    }

    public LiveData<Language> getTargetLang() {
        return targetLang;
    }

    private void translate(){
        if (originText.trim().isEmpty()){
            return;
        }
        if (disposable != null){
            disposable.dispose();
        }

        if (runTranslate != null){
            handler.removeCallbacks(runTranslate);
        }

        runTranslate = () -> {
            progress.setValue(true);
            failed.setValue(false);

            disposable = dataModel.translate(originLang.getValue(), targetLang.getValue(), originText)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(translateResponse -> {
                        translatedText.setValue(translateResponse.targetText);
                        progress.setValue(false);
                    }, throwable -> {
                        translatedText.setValue(throwable.getMessage());
                        progress.setValue(false);
                        failed.setValue(true);
                    });
        };

        handler.postDelayed(runTranslate, TRANSLATE_DELAY_MILLIS);
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        if (null != disposable){
            disposable.dispose();
        }
        if (runTranslate != null){
            handler.removeCallbacks(runTranslate);
        }
        handler = null;
    }
}

File: app/src/main/java/ru/santaev/clipboardtranslator/viewmodel/TranslateViewModel.kt
package ru.santaev.clipboardtranslator.viewmodel


import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import com.example.santaev.domain.dto.LanguageDto
import com.example.santaev.domain.factory.UseCaseFactory
import com.example.santaev.domain.usecase.TranslateUseCase
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import ru.santaev.clipboardtranslator.util.ILoggable
import ru.santaev.clipboardtranslator.util.RxHelper
import ru.santaev.clipboardtranslator.util.settings.AppPreference
import java.util.concurrent.TimeUnit

class TranslateViewModel : ViewModel(), ILoggable {

    val translatedText = MutableLiveData<String>()
    val progress = MutableLiveData<Boolean>()
    val failed = MutableLiveData<Boolean>()
    val sourceLanguage = MutableLiveData<LanguageDto?>()
    val targetLanguage = MutableLiveData<LanguageDto?>()
    var originText: String = ""
    private var translateDisposable: Disposable? = null
    private var languagesDisposable: Disposable? = null
    private val appPreference: AppPreference

    init {
        translatedText.value = ""
        progress.value = false
        failed.value = false
        appPreference = AppPreference.getInstance()

        val storedSourceLanguage = appPreference.originLang
        val storedTargetLanguage = appPreference.targetLang

        if (storedSourceLanguage == null || storedTargetLanguage == null) {
            loadLanguages()
        } else {
            sourceLanguage.value = storedSourceLanguage
            targetLanguage.value = storedTargetLanguage
        }
    }

    fun onOriginTextChanged(text: String) {
        originText = text
        translate()
    }

    fun onOriginLangSelected(lang: LanguageDto): Boolean {
        val oldLang = sourceLanguage.value
        sourceLanguage.value = lang
        appPreference.originLang = lang

        if (lang != oldLang) {
            translate()
        }
        return true
    }

    fun onTargetLangSelected(lang: LanguageDto): Boolean {
        val oldLang = targetLanguage.value
        targetLanguage.value = lang
        appPreference.targetLang = lang

        if (lang != oldLang) {
            translate()
        }
        return true
    }

    fun onClickRetry() {
        translate()
    }

    fun onClickSwipeLanguages() {
        val lang = targetLanguage.value
        targetLanguage.value = sourceLanguage.value
        sourceLanguage.value = lang

        appPreference.originLang = sourceLanguage.value
        appPreference.targetLang = targetLanguage.value

        translate()
    }

    override fun onCleared() {
        super.onCleared()
        translateDisposable?.dispose()
        languagesDisposable?.dispose()
    }

    private fun translate() {
        if (originText.trim { it <= ' ' }.isEmpty()) return

        val sourceLang = sourceLanguage.value ?: return
        val targetLang = targetLanguage.value ?: return

        progress.value = true
        failed.value = false

        translateDisposable?.dispose()

        translateDisposable = UseCaseFactory
                .instance
                .getTranslateFactory(
                        sourceLang,
                        targetLang,
                        originText
                )
                .execute()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        { response -> onTextTranslated(response) },
                        { onTranslateFailed() }
                )
    }

    private fun onTextTranslated(response: TranslateUseCase.Response) {
        when (response) {
            is TranslateUseCase.Response.Success -> {
                translatedText.value = response.targetText
                progress.value = false
            }
            is TranslateUseCase.Response.Error -> onTranslateFailed()
        }
    }

    private fun onTranslateFailed() {
        translatedText.value = "Error"
        progress.value = false
        failed.value = true
    }

    private fun loadLanguages() {
        languagesDisposable = UseCaseFactory
                .instance
                .getGetLanguagesUseCase()
                .execute()
                .filter { it.isNotEmpty() }
                .firstOrError()
                .compose(RxHelper.getSingleTransformer())
                .subscribe(
                        { languages -> setSourceAndTargetLanguages(languages) },
                        { error -> log("Unable to load languages", error) }
                )
    }

    private fun setSourceAndTargetLanguages(languages: List<LanguageDto>) {
        if (languages.isNotEmpty()) {
            val sourceLanguageDto = languages.firstOrNull { it.code == "en" } ?: languages.first()
            val targetLanguageDto = languages.firstOrNull { it.code == "ru" } ?: languages.first()
            sourceLanguage.value = sourceLanguageDto
            targetLanguage.value = targetLanguageDto
            appPreference.originLang = sourceLanguageDto
            appPreference.targetLang = targetLanguageDto
        }
    }

    companion object {

        private val TRANSLATE_DELAY_MILLIS = TimeUnit.MILLISECONDS.toMillis(500)
    }
}
